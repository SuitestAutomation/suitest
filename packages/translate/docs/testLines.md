# Test Lines

Translating test line definitions into human readable language.

## Prerequisites

In order to translate a line, you'll need the following data:

* Line definition, e.g. loaded from feed `/results/{resultId}`.
* Application configuration, e.g. from feed `/test-pack-runs/{testPackRunId}`, see `effectiveAppConfig` property.
    This information is needed because any configuration variables used in test need to be translated.
* Optional, a list of elements used during test execution, so that actual element names could be displayed instead of
    just ids. This information will be available from the feed `/results/{resultId}` shortly.
* Optional, a list of snippets used during test execution, so that actual snippet names could be displayed instead of
    just ids.

Reference the feeds can be found in [our documentation].

## Translating test line definition to a plain text

For example:
```javascript
const {testLineToPlainText} = require('@suitest/translate');

const translateLine = testLineToPlainText({type: 'sleep', timeout: 2000}, appConfiguration);

console.log(translateLine); // Sleep for 2s
```

## Translating test line definition to a formatted text

```javascript
const {testLineToFormattedText} = require('@suitest/translate');

const translateLine = testLineToFormattedText({
    type: 'assert',
    then: 'success',
    condition: {
        subject: {type: 'application'},
        type: 'exited'
    },
}, appConfiguration);

console.log(translatedLine); // Assert: Application has exited
```

## Translating test line definition to a HTML

```javascript
const {testLineToHtml} = require('@suitest/translate');

const translateLine = testLineToHtml({
    type: 'wait',
    then: 'success',
    timeout: 2000,
    condition: {
        subject: {type: 'cookie', val: 'my-cookie'},
        type: '=',
        val: 'test'
    },
}, appConfiguration);

console.log(translatedLine);
```

Output (in HTML):
```html
<figure class="suitest-test-line">
	<figcaption>Assert: Cookie</figcaption>
	<figure class="suitest-test-line__dictionary">
		<dl>
			<dt>Timeout</dt>
			<dd><span class="suitest-test-line__text--bold">2s</span></dd>
		</dl>
	</figure>
	<figure class="suitest-test-line__table">
		<table>
			<tr>
				<td><span class="suitest-test-line__text--bold">my-cookie</span> cookie</td>
				<td>=</td>
				<td><span class="suitest-test-line__text--bold">test</span></td>
			</tr>
		</table>
	</figure>
</figure>
```

In order to apply basic styling, make sure to include `@suitest/translate/suitest-test-line.css` file from
the NPM package into your HTML page.

## Other formats

We are using our own flavour of [unist] AST to translate test lines into multiple formats. Thus, you can implement
your own renderers if provided by default do not fit your needs.

```javascript
const {testLineToAst} = require('@suitest/translate');

const suitestAst = testLineToAst({type: 'sleep', timeout: 2000});

console.log(suitestAst);
```

Would produce:
```json5
{
  type: 'test-line',
  title: [
    {type: 'text', value: 'Sleep for '},
    {type: 'bold', value: '2'},
    {type: 'text', value: 's'},
  ],
}
```

[our documentation]: https://suite.st/docs/suitest-network-api/api-reference/
[unist]: https://github.com/syntax-tree/unist
