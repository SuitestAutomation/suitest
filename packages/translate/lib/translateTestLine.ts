import {AppConfiguration, Elements, Snippets, TestLine} from '@suitest/types';

import {testLineToAst as toAST} from './testLineToAst';
import {toText} from './renderers/text';
import {toHtml} from './renderers/html';
import {TestLineResult} from '@suitest/types/lib';

export const testLineToPlainText = ({
	lineDefinition, appConfig, elements, snippets, lineResult,
}: {
	lineDefinition: TestLine,
	appConfig: AppConfiguration,
	elements?: Elements,
	snippets?: Snippets,
	lineResult?: TestLineResult,
}): string => {
	const translation = toAST({testLine: lineDefinition, appConfig, elements, snippets, lineResult});

	return toText(translation, false);
};

export const testLineToFormattedText = ({
	lineDefinition, appConfig, elements, snippets, lineResult,
}: {
	lineDefinition: TestLine,
	appConfig: AppConfiguration,
	elements?: Elements,
	snippets?: Snippets,
	lineResult?: TestLineResult,
}): string => {
	const translation = toAST({testLine: lineDefinition, appConfig, elements, snippets, lineResult});

	return toText(translation, true);
};

export const testLineToHtml = ({
   lineDefinition, appConfig, elements, snippets, lineResult,
}: {
	lineDefinition: TestLine,
	appConfig: AppConfiguration,
	elements?: Elements,
	snippets?: Snippets,
	lineResult?: TestLineResult,
}): string => {
	const translation = toAST({testLine: lineDefinition, appConfig, elements, snippets, lineResult});

	return toHtml(translation);
};
