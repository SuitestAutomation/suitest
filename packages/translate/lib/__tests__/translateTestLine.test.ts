import {testLineToFormattedText, testLineToHtml, testLineToPlainText} from '../translateTestLine';
import {appConfig, testLines} from './testLines';

describe('translateTestLine module', () => {
	const line = testLines['Clear app data']();

	describe('testLineToPlainText function', () => {
		it('should connect AST builder function with plain text renderer', () => {
			expect(testLineToPlainText(line, appConfig)).toEqual('Clear application data\n');
		});
	});

	describe('testLineToFormattedText function', () => {
		it('should connect AST builder function with formatted text renderer', () => {
			expect(testLineToFormattedText(line, appConfig)).toEqual('Clear application data\n');
		});
	});

	describe('testLineToHtml function', () => {
		it('should connect AST builder function with HTML renderer', () => {
			expect(testLineToHtml(line, appConfig)).toMatchSnapshot();
		});
	});
});
