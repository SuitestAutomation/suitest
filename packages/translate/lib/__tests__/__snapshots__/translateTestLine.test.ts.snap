// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateTestLine module testLineToHtml function should connect AST builder function with HTML renderer 1`] = `"<figure class=\\"suitest-test-line\\"><figcaption>Clear application data</figcaption></figure>"`;
